lib/pdu.c

```diff
@@ -465,10 +500,36 @@ int rpc_queue_pdu(struct rpc_context *rpc, struct rpc_pdu *pdu)
         recordmarker = htonl(size | 0x80000000);
 	memcpy(pdu->out.iov[0].buf, &recordmarker, 4);
 
-	/* for udp we dont queue, we just send it straight away */
-	if (rpc->is_udp != 0) {
+	/*
+	 * For udp we dont queue, we just send it straight away.
+	 *
+	 * Another case where we send straight away is the AUTH_TLS NULL RPC.
+	 * This is particularly important for the reconnect case where we want to
+	 * ensure TLS handshake completes successfully before we can send any of
+	 * the queued RPCs waiting. If we do not send here this AUTH_TLS NULL
+	 * RPC will need to be queued before all other waiting RPCs and even then
+	 * we need to be careful that we don't send any of those RPCs till the
+	 * TLS handshake is completed and the connection is secure.
+	 * Sending inline here makes the handling simpler in rpc_service().
+	 */
+	if (rpc->is_udp != 0
```

Comment:
Personal style,
I would find
#ifdef HAVE_TLS
if (rpc->is_udp != 0 || pdu->expect_starttls) {
#else
if (rpc->is_udp != 0) {
#endif
easier to read.

Don't change it, it is fine. Just a personal comment.

Reply:
sure enough, I did have this question whether to do it this way or the way I did.
I chose the way I did as it's only slightly "less readable" if we may call it but it is less prone to both tls and non-tls code evolving differently. I'll leave this. Thanks.